# Dealing with Categorical Data

Many machine learning models can not deal with categorical data set as strings. For example linear regression can not apply a a Beta Coefficent to colors like "red" or "blue". Instead we need to convert these categories into "dummy" variables, otherwise known as "one-hot" encoding.

In [1]:
#import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
#read the final csv file with no missing data
df = pd.read_csv("../DATA/Ames_my_NO_Missing_Data.csv")

In [5]:
df.head()

MS SubClass MS Zoning  Lot Frontage  Lot Area Street Lot Shape  \
0           20        RL         141.0     31770   Pave       IR1   
1           20        RH          80.0     11622   Pave       Reg   
2           20        RL          81.0     14267   Pave       IR1   
3           20        RL          93.0     11160   Pave       Reg   
4           60        RL          74.0     13830   Pave       IR1   

  Land Contour Utilities Lot Config Land Slope  ... Enclosed Porch 3Ssn Porch  \
0          Lvl    AllPub     Corner        Gtl  ...              0          0   
1          Lvl    AllPub     Inside        Gtl  ...              0          0   
2          Lvl    AllPub     Corner        Gtl  ...              0          0   
3          Lvl    AllPub     Corner        Gtl  ...              0          0   
4          Lvl    AllPub     Inside        Gtl  ...              0          0   

  Screen Porch Pool Area Misc Val  Mo Sold  Yr Sold  Sale Type  \
0            0         0        0        5     2010        WD    
1          120         0        0        6     2010        WD    
2            0         0    12500        6     2010        WD    
3            0         0        0        4     2010        WD    
4            0         0        0        3     2010        WD    

   Sale Condition SalePrice  
0          Normal    215000  
1          Normal    105000  
2          Normal    172000  
3          Normal    244000  
4          Normal    189900  

[5 rows x 76 columns]

## Numerical Column to Categorical

We need to be careful when it comes to encoding categories as numbers. We want to make sure that the numerical relationship makes sense for a model. For example, the encoding MSSubClass is essentially just a number code per class:

    MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

The number itself does not appear to have a relationship to the other numbers. While 30 > 20 is True, it doesn't really make sense that "1-STORY 1945 & OLDER" > "1-STORY 1946 & NEWER ALL STYLES". Keep in mind, this isn't always the case, for example 1st class seats versus 2nd class seats encoded as 1 and 2. Make sure you fully understand your data set to examine what needs to be converted/changed.

### MSSubClass

In [6]:
# Convert to String
df['MS SubClass'] = df['MS SubClass'].apply(str)

## Creating Dummy Variables from Object Columns

In [7]:
df.select_dtypes(include='object')

MS SubClass MS Zoning Street Lot Shape Land Contour Utilities Lot Config  \
0             20        RL   Pave       IR1          Lvl    AllPub     Corner   
1             20        RH   Pave       Reg          Lvl    AllPub     Inside   
2             20        RL   Pave       IR1          Lvl    AllPub     Corner   
3             20        RL   Pave       Reg          Lvl    AllPub     Corner   
4             60        RL   Pave       IR1          Lvl    AllPub     Inside   
...          ...       ...    ...       ...          ...       ...        ...   
2920          80        RL   Pave       IR1          Lvl    AllPub    CulDSac   
2921          20        RL   Pave       IR1          Low    AllPub     Inside   
2922          85        RL   Pave       Reg          Lvl    AllPub     Inside   
2923          20        RL   Pave       Reg          Lvl    AllPub     Inside   
2924          60        RL   Pave       Reg          Lvl    AllPub     Inside   

     Land Slope Neighborhood Condition 1  ... Kitchen Qual Functional  \
0           Gtl        NAmes        Norm  ...           TA        Typ   
1           Gtl        NAmes       Feedr  ...           TA        Typ   
2           Gtl        NAmes        Norm  ...           Gd        Typ   
3           Gtl        NAmes        Norm  ...           Ex        Typ   
4           Gtl      Gilbert        Norm  ...           TA        Typ   
...         ...          ...         ...  ...          ...        ...   
2920        Gtl      Mitchel        Norm  ...           TA        Typ   
2921        Mod      Mitchel        Norm  ...           TA        Typ   
2922        Gtl      Mitchel        Norm  ...           TA        Typ   
2923        Mod      Mitchel        Norm  ...           TA        Typ   
2924        Mod      Mitchel        Norm  ...           TA        Typ   

     Fireplace Qu Garage Type Garage Finish Garage Qual Garage Cond  \
0              Gd      Attchd           Fin          TA          TA   
1            None      Attchd           Unf          TA          TA   
2            None      Attchd           Unf          TA          TA   
3              TA      Attchd           Fin          TA          TA   
4              TA      Attchd           Fin          TA          TA   
...           ...         ...           ...         ...         ...   
2920         None      Detchd           Unf          TA          TA   
2921         None      Attchd           Unf          TA          TA   
2922         None        None          None        None        None   
2923           TA      Attchd           RFn          TA          TA   
2924           TA      Attchd           Fin          TA          TA   

     Paved Drive Sale Type Sale Condition  
0              P       WD          Normal  
1              Y       WD          Normal  
2              Y       WD          Normal  
3              Y       WD          Normal  
4              Y       WD          Normal  
...          ...       ...            ...  
2920           Y       WD          Normal  
2921           Y       WD          Normal  
2922           Y       WD          Normal  
2923           Y       WD          Normal  
2924           Y       WD          Normal  

[2925 rows x 40 columns]

In [8]:
df_nums = df.select_dtypes(exclude='object')
df_objs = df.select_dtypes(include='object')

In [9]:
df_nums.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925 entries, 0 to 2924
Data columns (total 36 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Lot Frontage     2925 non-null   float64
 1   Lot Area         2925 non-null   int64  
 2   Overall Qual     2925 non-null   int64  
 3   Overall Cond     2925 non-null   int64  
 4   Year Built       2925 non-null   int64  
 5   Year Remod/Add   2925 non-null   int64  
 6   Mas Vnr Area     2925 non-null   float64
 7   BsmtFin SF 1     2925 non-null   float64
 8   BsmtFin SF 2     2925 non-null   float64
 9   Bsmt Unf SF      2925 non-null   float64
 10  Total Bsmt SF    2925 non-null   float64
 11  1st Flr SF       2925 non-null   int64  
 12  2nd Flr SF       2925 non-null   int64  
 13  Low Qual Fin SF  2925 non-null   int64  
 14  Gr Liv Area      2925 non-null   int64  
 15  Bsmt Full Bath   2925 non-null   float64
 16  Bsmt Half Bath   2925 non-null   float64
 17  Full Bath     

In [10]:
df_objs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925 entries, 0 to 2924
Data columns (total 40 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   MS SubClass     2925 non-null   object
 1   MS Zoning       2925 non-null   object
 2   Street          2925 non-null   object
 3   Lot Shape       2925 non-null   object
 4   Land Contour    2925 non-null   object
 5   Utilities       2925 non-null   object
 6   Lot Config      2925 non-null   object
 7   Land Slope      2925 non-null   object
 8   Neighborhood    2925 non-null   object
 9   Condition 1     2925 non-null   object
 10  Condition 2     2925 non-null   object
 11  Bldg Type       2925 non-null   object
 12  House Style     2925 non-null   object
 13  Roof Style      2925 non-null   object
 14  Roof Matl       2925 non-null   object
 15  Exterior 1st    2925 non-null   object
 16  Exterior 2nd    2925 non-null   object
 17  Mas Vnr Type    2925 non-null   object
 18  Exter Qu

### Converting the object df 

In [11]:
df_objs = pd.get_dummies(df_objs,drop_first=True)

In [12]:
final_df = pd.concat([df_nums,df_objs],axis=1)

In [13]:
final_df

Lot Frontage  Lot Area  Overall Qual  Overall Cond  Year Built  \
0       141.000000     31770             6             5        1960   
1        80.000000     11622             5             6        1961   
2        81.000000     14267             6             6        1958   
3        93.000000     11160             7             5        1968   
4        74.000000     13830             5             5        1997   
...            ...       ...           ...           ...         ...   
2920     37.000000      7937             6             6        1984   
2921     75.144444      8885             5             5        1983   
2922     62.000000     10441             5             5        1992   
2923     77.000000     10010             5             5        1974   
2924     74.000000      9627             7             5        1993   

      Year Remod/Add  Mas Vnr Area  BsmtFin SF 1  BsmtFin SF 2  Bsmt Unf SF  \
0               1960         112.0         639.0           0.0        441.0   
1               1961           0.0         468.0         144.0        270.0   
2               1958         108.0         923.0           0.0        406.0   
3               1968           0.0        1065.0           0.0       1045.0   
4               1998           0.0         791.0           0.0        137.0   
...              ...           ...           ...           ...          ...   
2920            1984           0.0         819.0           0.0        184.0   
2921            1983           0.0         301.0         324.0        239.0   
2922            1992           0.0         337.0           0.0        575.0   
2923            1975           0.0        1071.0         123.0        195.0   
2924            1994          94.0         758.0           0.0        238.0   

      ...  Sale Type_ConLw  Sale Type_New  Sale Type_Oth  Sale Type_VWD  \
0     ...                0              0              0              0   
1     ...                0              0              0              0   
2     ...                0              0              0              0   
3     ...                0              0              0              0   
4     ...                0              0              0              0   
...   ...              ...            ...            ...            ...   
2920  ...                0              0              0              0   
2921  ...                0              0              0              0   
2922  ...                0              0              0              0   
2923  ...                0              0              0              0   
2924  ...                0              0              0              0   

      Sale Type_WD   Sale Condition_AdjLand  Sale Condition_Alloca  \
0                 1                       0                      0   
1                 1                       0                      0   
2                 1                       0                      0   
3                 1                       0                      0   
4                 1                       0                      0   
...             ...                     ...                    ...   
2920              1                       0                      0   
2921              1                       0                      0   
2922              1                       0                      0   
2923              1                       0                      0   
2924              1                       0                      0   

      Sale Condition_Family  Sale Condition_Normal  Sale Condition_Partial  
0                         0                      1                       0  
1                         0                      1                       0  
2                         0                      1                       0  
3                         0                      1                       0  
4                         0                      1               

In [14]:
final_df.corr()['SalePrice'].sort_values()

Exter Qual_TA       -0.591459
Kitchen Qual_TA     -0.527461
Fireplace Qu_None   -0.481740
Bsmt Qual_TA        -0.453022
Garage Finish_Unf   -0.422363
                       ...   
Garage Cars          0.648488
Total Bsmt SF        0.660983
Gr Liv Area          0.727279
Overall Qual         0.802637
SalePrice            1.000000
Name: SalePrice, Length: 274, dtype: float64

## Final Thoughts

Keep in mind, we don't know if 274 columns is very useful. More columns doesn't necessarily lead to better results. In fact, we may want to further remove columns (or later on use a model with regularization to choose important columns for us). What we have done here has greatly expanded the ratio of rows to columns, which may actually lead to worse performance (however you don't know until you've actually compared multiple models/approaches).

In [16]:
#save the final data frame as a csv file to use it in the model building part
final_df.to_csv('../DATA/AMES_Final_DF.csv')